### PR TITLE
[fix #509] add i18n typed check

### LINKED ISSUE #509: Localization JSONs: scan the source code to create a complete list of all locale keys, and verify master English locale (no missing, no redundant keys)
This is an adjacent issue to https://github.com/readium/readium-desktop/issues/397

### PR DESCRIPTION
**Pull request by panaC**
I created a i18n typing with a tools https://github.com/kogai/typed_i18n/issues

It's allow to check if a key is typed. automatic typing with vscode on __("mykey");

To regenerate the typing file: `npm run i18n-typed`

It's doesn't allow to check redundant keys.

fix #509

be careful to generate the typing file with `npm run i18n-typed` at each update of json locale file

### CODE REVIEW

**Comment by danielweck** [comment]
Sophisticated approach, I like it :)
(reviewing...)

File: src/common/services/translator.ts [hunk 1]
```diff
@@ -15,20 +15,11 @@ import * as frCatalog from "readium-desktop/resources/locales/fr.json";
 import * as deLang from "readium-desktop/resources/locale-names/enLang.json";
 import * as enLang from "readium-desktop/resources/locale-names/enLang.json";
 import * as frLang from "readium-desktop/resources/locale-names/frLang.json";
+import { TFunction } from "readium-desktop/typings/en.translation";
 
 const initI18n = i18n.init({
+    resources: Object.assign(enCatalog, deCatalog, frCatalog),
```
Comment (danielweck) [inline src/common/services/translator.ts]: I have a request: by design, the various JSON locale files are currently (i.e. `develop` branch) comparable in their canonical form, for example key sorting + value removal/ignoring. This makes it trivial order to detect missing or redundant keys across multiple languages, using off the shelf tooling (as implemented in the `develop` branch). So, although moving the `LANG_CODE.translation` structure into each locale JSON is elegant, this also breaks existing tooling. Is it possible to restore this feature? Or is this change essential to the `I18nTyped` implementation?

**Comment by danielweck** [comment]
I do like the proposed technique which introduces robust "typing" of locale JSONs.

Questions:

* Does the technique implemented in this PR help us identify missing keys in existing `locales/LANG.json` files?
* Same question for redundant/unused keys.

Either way, I would like to adopt the PR, which introduces robustness using a sophisticated technique.
However, this can be combined with the lighweight technique implemented in this PR (which solves both questions above): https://github.com/readium/readium-desktop/pull/516

**Comment by panaC** [comment]
can you remove the *.d.ts linting rules ? or Why vscode isn't incrementing correctly ?

File: src/common/services/translator.ts (hunk 1 above)
Comment (panaC) [reply src/common/services/translator.ts]: this change is essential to the I18nTyped implementation, sorry. It's the best practice https://www.i18next.com/principles/translation-resolution. https://github.com/kogai/typed_i18n/issues/66#issuecomment-456303425

Comment (danielweck) [reply src/common/services/translator.ts]: Ok, we will work around this breaking change, no problems.

**Comment by danielweck** [comment]
So, this new "strong locale typing" feature is a fantastic helper, but in practice this has a prerequisite: that the `en` primary language JSON is clean (no missing keys, no redundant keys). Which is why ; I think ; we need the existing tooling in place, plus the pending PR. The JSON structure refactoring needs to be done first.

**Comment by danielweck** [comment]
Follow-up issue: https://github.com/readium/readium-desktop/issues/514

**Pull request merged**